rewriter/Minitest.cc

```diff
@@ -220,19 +239,45 @@ ast::ExpressionPtr runUnderEach(core::MutableContext ctx, core::NameRef eachName
     // this statement must be a send
     if (auto send = ast::cast_tree<ast::Send>(stmt)) {
         auto correctBlockArity = send->hasBlock() && send->block()->args.size() == 0;
-        // the send must be a call to `it` with a single argument (the test name) and a block with no arguments
-        if ((send->fun == core::Names::it() && send->numPosArgs() == 1 && correctBlockArity) ||
-            ((send->fun == core::Names::before() || send->fun == core::Names::after()) && send->numPosArgs() == 0 &&
-             correctBlockArity)) {
+
+        // Check for different types of test blocks
+        auto isItOrXit = (send->fun == core::Names::it() || send->fun == core::Names::xit());
+        auto isIts = (send->fun == core::Names::its());
+        auto isBeforeOrAfter = (send->fun == core::Names::before() || send->fun == core::Names::after());
+
+        auto itOrXitWithValidArgs = isItOrXit && (send->numPosArgs() == 0 || send->numPosArgs() == 1);
+        auto itsWithValidArgs = isIts && send->numPosArgs() == 1;
+        auto beforeOrAfterWithValidArgs = isBeforeOrAfter && send->numPosArgs() == 0;
+
+        auto isValidTestBlock = (itOrXitWithValidArgs || itsWithValidArgs) && correctBlockArity;
+        auto isValidSetupTeardown = beforeOrAfterWithValidArgs && correctBlockArity;
+
+        if (isValidTestBlock || isValidSetupTeardown) {
             core::NameRef name;
             if (send->fun == core::Names::before()) {
                 name = core::Names::beforeAngles();
             } else if (send->fun == core::Names::after()) {
                 name = core::Names::afterAngles();
             } else {
                 // we use this for the name of our test
-                auto argString = to_s(ctx, send->getPosArg(0));
-                name = ctx.state.enterNameUTF8("<it '" + argString + "'>");
+                if (send->numPosArgs() == 1) {
+                    auto argString = to_s(ctx, send->getPosArg(0));
+                    if (send->fun == core::Names::its()) {
+                        name = ctx.state.enterNameUTF8("<its " + argString + ">");
+                    } else if (send->fun == core::Names::xit()) {
+                        name = ctx.state.enterNameUTF8("<xit '" + argString + "'>");
+                    } else {
+                        name = ctx.state.enterNameUTF8("<it '" + argString + "'>");
+                    }
+                } else {
+                    // no argument provided, use a deterministic unique name based on location
+                    auto pos = send->loc.beginPos();
+                    if (send->fun == core::Names::xit()) {
+                        name = ctx.state.enterNameUTF8("<anonymous_xit_" + to_string(pos) + ">");
+                    } else {
+                        name = ctx.state.enterNameUTF8("<anonymous_it_" + to_string(pos) + ">");
+                    }
```

Comment:
Can we get away with all of these methods just having no name?


```suggestion
                    if (send->fun == core::Names::xit()) {
                        name = core::Names::anonymousXit();
                    } else {
                        name = core::Names::anonymousIt();
                    }
```

I don't think that anything depends on the names of these methods, and there's no problem with having multiple methods with the same name (e.g., we're not doing anything to prevent you from writing a named test with a name that was already used, and that doesn't cause problems).

Reply:
The context is that I don't want AST contents to be a function of things like how much whitespace there is in the file.

Reply:
Good point. I think I was under some assumption that the names had to be unique... fixed here fcc5721ab